test/Sema/struct_equatable_hashable.swift

```diff
@@ -0,0 +1,154 @@
+
+
+// Structs with non-hashable/equatable stored properties don't derive conformance.
+struct NotHashable {}
+struct StructWithNonHashablePayload: Hashable { // expected-error 2 {{does not conform}}
+  let a: NotHashable
+}
+
+// ...but computed properties and static properties are not considered.
+struct StructIgnoresComputedProperties: Hashable {
+  var a: Int
+  var b: String
+  static var staticComputed = NotHashable()
+  var computed: NotHashable { return NotHashable() }
+}
+if StructIgnoresComputedProperties(a: 1, b: "a") == StructIgnoresComputedProperties(a: 2, b: "c") {}
+let _: Int = StructIgnoresComputedProperties(a: 3, b: "p").hashValue
+
+
+// Structs should be able to derive conformances based on the conformances of
+// their generic arguments.
+struct GenericHashable<T: Hashable>: Hashable {
+  let value: T
+}
+if GenericHashable<String>(value: "a") == GenericHashable<String>(value: "b") { }
+var genericHashableHash: Int = GenericHashable<String>(value: "a").hashValue
+
+// But it should be an error if the generic argument doesn't have the necessary
+// constraints to satisfy the conditions for derivation.
+struct GenericNotHashable<T: Equatable>: Hashable { // expected-error {{does not conform}}
+  let value: T
+}
+if GenericNotHashable<String>(value: "a") == GenericNotHashable<String>(value: "b") { }
+var genericNotHashableHash: Int = GenericNotHashable<String>(value: "a").hashValue // expected-error {{value of type 'GenericNotHashable<String>' has no member 'hashValue'}}
+
+
+// Conformance cannot be synthesized in an extension.
+struct StructConformsInExtension {
+  let v: Int
+}
+extension StructConformsInExtension : Equatable {} // expected-error {{cannot be automatically synthesized in an extension}}
+
+// But explicit conformance in an extension should work.
+public struct StructConformsAndImplementsInExtension {
+  let v: Int
+}
+extension StructConformsAndImplementsInExtension : Equatable {
+  public static func ==(lhs: StructConformsAndImplementsInExtension, rhs: StructConformsAndImplementsInExtension) -> Bool {  // expected-note {{non-matching type}}
+    return true
+  }  
+}
+
+// No explicit conformance and it cannot be derived.
+struct NotExplicitlyHashableAndCannotDerive {
+  let v: NotHashable
+}
+extension NotExplicitlyHashableAndCannotDerive : Hashable {}  // expected-error 2 {{does not conform}}
+
+// A struct with no stored properties trivially derives conformance.
+struct NoStoredProperties: Hashable {}
+
+// Verify that conformance (albeit manually implemented) can still be added to
+// a type in a different file.
+extension OtherFileNonconforming: Hashable {
+  static func ==(lhs: OtherFileNonconforming, rhs: OtherFileNonconforming) -> Bool { // expected-note {{non-matching type}}
+    return true
+  }
+  var hashValue: Int { return 0 }
+}
+// ...but synthesis in a type defined in another file doesn't work yet.
```

Comment:
Can you file a bug to investigate this if you don't plan on doing it in this PR?

Reply:
Filed [SR-6101](https://bugs.swift.org/browse/SR-6101).